Mathlib/Order/RelSeries.lean

```diff
@@ -154,6 +157,124 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
   ⟨RelSeries.withLength r 0 0⟩
 
+instance membership : Membership α (RelSeries r) :=
+  ⟨(· ∈ Set.range ·)⟩
+
+theorem mem_def {x : α} {s : RelSeries r} : x ∈ s ↔ x ∈ Set.range s :=
+  Iff.rfl
+
+/-- Start of a series, i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its head is `a₀`. Since a relation
+series is assumed to be non-empty, this is well defined. -/
+def head (x : RelSeries r) : α := x 0
+
+/-- End of a series, i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its last element is `aₙ`.  Since a
+relation series is assumed to be non-empty, this is well defined. -/
+def last (x : RelSeries r) : α := x <| Fin.last _
+
+lemma head_mem (x : RelSeries r) : x.head ∈ x := ⟨_, rfl⟩
+
+lemma last_mem (x : RelSeries r) : x.last ∈ x := ⟨_, rfl⟩
+
+/--
+If `a_0 --r-> a_1 --r-> ... --r-> a_n` and `b_0 --r-> b_1 --r-> ... --r-> b_m` are two strict series
+such that `r a_n b_0`, then there is a chain of length `n + m + 1` given by
+`a_0 --r-> a_1 --r-> ... --r-> a_n --r-> b_0 --r-> b_1 --r-> ... --r-> b_m`.
+-/
+@[simps]
+def append (p q : RelSeries r) (connect : r p.last q.head) : RelSeries r where
+  length := p.length + q.length + 1
+  toFun := Fin.append p q ∘ Fin.cast (by abel)
+  step := fun i => by
+    obtain (hi|rfl|hi) :=
+      lt_trichotomy i (Fin.castLE (by linarith) (Fin.last _ : Fin (p.length + 1)))
+    · convert p.step ⟨i.1, hi⟩ <;> convert Fin.append_left p q _ <;> rfl
+    · convert connect
+      · convert Fin.append_left p q _; rfl
+      · convert Fin.append_right p q _; rfl
+    · set x := _; set y := _
+      change r (Fin.append p q x) (Fin.append p q y)
+      have hx : x = Fin.natAdd _ ⟨i - (p.length + 1), Nat.sub_lt_left_of_lt_add hi <|
+        i.2.trans <| by linarith!⟩
+      · ext; dsimp; rw [Nat.add_sub_cancel']; exact hi
+      have hy : y = Fin.natAdd _ ⟨i - p.length, Nat.sub_lt_left_of_lt_add (le_of_lt hi)
+        (by exact i.2)⟩
+      · ext
+        dsimp
+        conv_rhs => rw [Nat.add_comm p.length 1, add_assoc,
+          Nat.add_sub_cancel' <| le_of_lt (show p.length < i.1 from hi), add_comm]
+      rw [hx, Fin.append_right, hy, Fin.append_right]
+      convert q.step ⟨i - (p.length + 1), Nat.sub_lt_left_of_lt_add hi <|
+        by convert i.2 using 1; abel⟩
+      rw [Fin.succ_mk, Nat.sub_eq_iff_eq_add (le_of_lt hi : p.length ≤ i),
+        Nat.add_assoc _ 1, add_comm 1, Nat.sub_add_cancel]
+      exact hi
+
+/--
+For two sets `α, β` and relation on them `r, s`, if `f : α → β` preserves relation `r`, then an
+`r`-series can be pushed out to an `s`-series by
+`a₀ --r-> a₁ --r-> ... --r-> aₙ ↦ f a₀ --s-> f a₁ --s-> ... --s-> f aₙ`
+-/
+@[simps]
+def map (p : RelSeries r)
+    (f : α → β) (hf : ∀ ⦃x y : α⦄, r x y → s (f x) (f y)) : RelSeries s where
+  length := p.length
+  toFun := f.comp p
+  step := (hf <| p.step .)
+
+/--
+If `a_0 --r-> a_1 --r-> ... --r-> a_n` is an `r`-series and `a` is such that
+`a_i --r-> a --r-> a_{i + 1}`, then
+`a_0 --r-> a_1 --r-> ... --r-> a_i --r-> a --r-> a_{i + 1} --r-> ... --r-> a_n`
```

Comment:
This notation is hard to parse. Can you either
* Use Unicode subscripts:
```suggestion
If `a₀ -r→ a₁ -r→ ... -r→ aₙ` is an `r`-series and `a` is such that
`aᵢ -r→ a -r→ a_ᵢ₊₁`, then
`a₀ -r→ a₁ -r→ ... -r→ a_i -r→ a -r→ aᵢ₊₁ -r→ ... -r→ aₙ`
```
or use latex?